centml/compiler/backend.py

```diff
@@ -1,18 +1,145 @@
+
+    def __compile_model(self, model_id: str):
+        compile_response = requests.post(
+            url=f"{server_url}/submit/{model_id}",
+            files={"model": pickle.dumps(self.module), "inputs": pickle.dumps(self.inputs)},
+            timeout=config_instance.TIMEOUT,
+        )
+        return compile_response
+
+    def __wait_for_status(self, model_id: str) -> bool:
+        tries = 0
+        while True:
+            # get server compilation status
+            status_response = requests.get(f"{server_url}/status/{model_id}", timeout=config_instance.TIMEOUT)
+            if status_response.status_code != HTTPStatus.OK:
+                self.exception_logs += f"Status check: exception from server: {status_response.json().get('detail')}"
+                return False
+            status = status_response.json().get("status")
+
+            if status == CompilationStatus.DONE.value:
+                return True
+            elif status == CompilationStatus.COMPILING.value:
+                pass
+            elif status == CompilationStatus.NOT_FOUND.value:
+                tries += 1
+                compiled_response = self.__compile_model(model_id)
+                if compiled_response.status_code != HTTPStatus.OK:
+                    self.exception_logs += f"{compiled_response.json().get('detail')}\n"
+            else:
+                tries += 1
+
+            if tries > config_instance.MAX_RETRIES:
+                self.exception_logs += "Waiting for status: compilation failed too many times.\n"
+                return False
+
+            time.sleep(config_instance.COMPILING_SLEEP_TIME)
+
+    def remote_compilation(self):
+        # start by getting the model_id
+        interpreter: Interpreter = hidet.frontend.from_torch(self.module)
+        flow_graph, inputs, output_format = get_flow_graph(interpreter, self.inputs)
+        model_id = self.__get_model_id(flow_graph)
+        if model_id is None:
+            self.__print_exceptions()
+            return
+
+        cgraph_path = os.path.join(base_path, model_id, "cgraph.zip")
+        if os.path.isfile(cgraph_path):  # cgraph is saved locally
+            cgraph = load_compiled_graph(cgraph_path)
+        else:
+            if not self.__wait_for_status(model_id):
+                self.__print_exceptions()
+                return
+
+            cgraph = self.__download_model(model_id)
+            if cgraph is None:
+                self.__print_exceptions()
+                return
+
+        wrapper = get_wrapper(cgraph, inputs, output_format)
+        self.compiled_forward_function = wrapper
+
     def __call__(self, *args, **kwargs):
-        return self.module.forward(*args, **kwargs)
+        # If model is currently compiling, return the uncompiled forward function
+        if not self.compiled_forward_function:
+            return self.module(*args, **kwargs)
+
+        forward_function = self.compiled_forward_function(*args)
+        return forward_function
```

Comment:
you're right. We're not returning the forward function itself, but the result of calling the forward function. I can just return the result directly instead of saving it in a variable first